File: src/test/java/co/nlighten/jsontransform/functions/TransformerFunctionFilterTest.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.BaseTest;
import org.junit.jupiter.api.Test;

public class TransformerFunctionFilterTest extends BaseTest {
    @Test
    void truthyValuesOnly() {
        assertTransformation(fromJson("""
["a", true, "false", 0, 1, [], [0]]"""), fromJson("""
{
  "$$filter": "$", "by": "##current"
}
"""), fromJson("""
["a", true, "false", 1, [0]]"""));

        // explicit boolean (non js style)
        assertTransformation(fromJson("""
["a", true, "false", 0, 1, [], [0]]"""), fromJson("""
{
  "$$filter": "$", "by": "$$boolean:##current"
}
"""), fromJson("""
[true, 1, [0]]"""));
    }
    @Test
    void namesThatStartsWithA() {
        assertTransformation(fromJson("""
[{"name":"alice"}, {"name":"ann"}, {"name":"bob"}]"""), fromJson("""
{
  "$$filter": "$", "by": "$$test(^a):##current.name"
}
"""), fromJson("""
[{"name":"alice"}, {"name":"ann"}]"""));
    }

    @Test
    void inline() {
        assertTransformation(fromJson("[\"a\", true, \"false\", 0, 1, [], [0]]"),
            "$$filter(##current):$",
            fromJson("[\"a\", true, \"false\", 1, [0]]")
        );
    }
}

File: javascript/json-transform/src/__tests__/functions/TransformerFunctionFilter.test.ts
import { describe, test } from "vitest";
import { assertTransformation } from "../BaseTransformationTest";

describe("TransformerFunctionFilter", () => {
  test("truthyValuesOnly", async () => {
    await assertTransformation(
      ["a", true, "false", 0, 1, [], [0]],
      {
        $$filter: "$",
        by: "##current",
      },
      ["a", true, "false", 1, [0]],
    );

    // explicit boolean (non js style)
    await assertTransformation(
      ["a", true, "false", 0, 1, [], [0]],
      {
        $$filter: "$",
        by: "$$boolean:##current",
      },
      [true, 1, [0]],
    );
  });
  test("namesThatStartsWithA", async () => {
    await assertTransformation(
      [{ name: "alice" }, { name: "ann" }, { name: "bob" }],
      {
        $$filter: "$",
        by: "$$test(^a):##current.name",
      },
      [{ name: "alice" }, { name: "ann" }],
    );
  });

  test("inline", async () => {
    const expected = ["a", true, "false", 1, [0]];
    await assertTransformation(["a", true, "false", 0, 1, [], [0]], "$$filter(##current):$", expected);
  });
});
